lib/Onyx.d.ts

```diff
@@ -0,0 +1,278 @@
+import {Component} from 'react';
+import {PartialDeep} from 'type-fest';
+import * as Logger from './Logger';
+import {CollectionKey, DeepRecord, Key, OnyxKey, KeyValueMapping, Selector} from './types';
+
+type NullableKeyValueMapping = {
+    [TKey in OnyxKey]: KeyValueMapping[TKey] | null;
+};
+
+type BaseConnectOptions<TKey extends OnyxKey> = {
+    statePropertyName?: string;
+    withOnyxInstance?: Component;
+    initWithStoredValues?: boolean;
+    selector?: Selector<TKey, KeyValueMapping[TKey] | null>;
+};
+
+type ConnectOptions<TKey extends OnyxKey> = BaseConnectOptions<TKey> &
+    (
+        | {
+              key: TKey extends CollectionKey ? TKey : never;
+              callback?: (value: Record<string, KeyValueMapping[TKey] | null> | null, key?: TKey) => void;
+              waitForCollectionCallback: true;
+          }
+        | {
+              key: TKey;
+              callback?: (value: KeyValueMapping[TKey] | null, key?: TKey) => void;
+              waitForCollectionCallback?: false;
+          }
+    );
+
+type Collection<TKey extends Key, TMap, TValue> = {
+    [MapK in keyof TMap]: MapK extends `${TKey}${string}`
+        ? MapK extends `${TKey}`
+            ? never // forbids empty id
+            : TValue
+        : never;
+};
+
+type OnyxUpdate<TKey extends OnyxKey> =
+    | {
+          onyxMethod: typeof METHOD.SET;
+          key: TKey;
+          value: KeyValueMapping[TKey] | null;
+      }
+    | {
+          onyxMethod: typeof METHOD.MERGE;
+          key: TKey;
+          value: PartialDeep<KeyValueMapping[TKey]>;
+      }
+    | {
+          [TKey in CollectionKey]: {
+              onyxMethod: typeof METHOD.MERGE_COLLECTION;
+              key: TKey;
+              value: Record<`${TKey}${string}`, PartialDeep<KeyValueMapping[TKey]>>;
+          };
+      }[CollectionKey];
+
+type OnyxUpdates<TKeyList extends OnyxKey[]> = {
+    [TKey in keyof TKeyList]: OnyxUpdate<TKeyList[TKey]>;
+};
```

Comment:
I still have a problem with this declaration. It's declaring that OnyxUpdates is a mapped type that should result in an object-like type, not an array type. Onyx updates is an array of Onyx Updates. Please update.

Maybe this works?
```ts
type OnyxUpdates<TKeyList extends OnyxKey[]> = Array<
    {
        [TKey in keyof TKeyList]: OnyxUpdate<TKeyList[TKey]>;
    }[number]
>;
```
That way we can support this format for OnyxUpdates?
```ts
const onyxUpdates: OnyxUpdates<['testKey']> = [
    {
        onyxMethod: METHOD.SET,
        key: 'testKey',
        value: 'testString',
    },
];

```

Reply:
@neil-marcellini This might be a little bit confusing, but `OnyxUpdates` actually doesn't result in an object type but in an array type.

The key part is `keyof` in `[TKey in keyof TKeyList]`

TKey is actually an array index. So we're creating an object like this

```ts
{
 0: OnyxUpdate<TKeyList[0]>,
 1: OnyxUpdate<TKeyList[1]>,
 2: OnyxUpdate<TKeyList[2]>,
}
```

which is actually

```
[OnyxUpdate<TKeyList[0]>, OnyxUpdate<TKeyList[1]>, OnyxUpdate<TKeyList[2]>`
```

Reply:
Thanks for the write up Hayata. I would like to remove any type checking that doesn't work in the case described above, since it would be needless complexity.

Reply:
@hayata-suenaga The `update` function signature in your example is wrong, [here is an example with the correct one](https://www.typescriptlang.org/play?ts=4.8.4#code/C4TwDgpgBA0hJQLxQOQGt4EYVQD6oxACYd914BmFAbgChRIoA1AQwBsBXaZAb1qgEEsOAFxQAzsABOASwB2AczqChxFGIBGAey1sILOcsHkQVMXI4BbDRCl0AvrXrhoAVTAATFsAgAeACpQEAAePnIe4rDwAHxIUHwqhGL+ADT8ggBu7FxirJwQANr+ALq0jgxunt4Q4gFwIAAyMpJBoRDhkfUFxbG86QIF9VDyUIRaAGZQ-vVNksVi7l4+dfCzwIPwPQ5OFVGNzcBxBSbYKapUxXS7AEo1HGyHyIvVtTMH0XS0HhAAxmwsUmg4w4ch+wBkWjkUA4VWW01WB1aYQie260QAFDCljUFrCar4CgA6Ynw-ZzaIASjEGS0Mg8nyx1XRBR4SVUpygWXyYhQklkihQ9jOrPgPMIVDOXJyqD58gUguKFOoQA).

In this way we don't need to pass `TKeyList` as it will be inferred automatically by the function.

> There is no way to know for which keys updates are pushed from the backend

How are we going to type the data from backend? I'm assuming that we are going to use something like this.

```ts
const jsonData: any[] = [...]; // data from backend, we know that it is an array but we don't have control over the contents.
Onyx.update(jsonData); // this works.
```

If you are ok with that we don't need to remove the type-safety from `update`.

cc @neil-marcellini @blazejkustra

Reply:
ah I forgot to do `[...TKeyList]`. nice idea to use tuple 👍  and I apologize for my mistake 🙇 

yea in that case, let's keep the code as is.

Although we probably have to use `any` for the data received from the backend as we have to assert the type (because we don't know the type of Key passed to OnyxUpdate)

```
const jsonData: unkown[] = [...]; // data from backend, we know that it is an array but we don't have control over the contents.
Onyx.update(jsonData); // this works.
```
<img width="300" alt="Screenshot 2023-07-27 at 11 09 34 AM" src="https://github.com/Expensify/react-native-onyx/assets/98560306/53d9a73b-88f8-4c49-96e8-63b

Reply:
I'm coming back to this to make sure I understand where we landed. The data coming from the backend that we pass into Onyx.update will be an array of OnyxUpdates which is potentially empty. OnyxUpdates have `method: string, key: string, and value: Record<string, unknown> | null`. The current implementation can handle that correct?

Reply:
@hayata-suenaga @neil-marcellini I found a new way to define `OnyxUpdate`! With this way we are going to be able to use `OnyxUpdate` type with an array and use it in both `update()` method and in the App (e.g. when defining update arrays like `optimisticData`, `succesData`, `failureData`).

[Here is the TS Playground.](https://www.typescriptlang.org/play?ts=4.8.4#code/C4TwDgpgBAaghgGwK4QPIDMA8qBGArCAY2ABVwIAaWRFMyKCAD2AgDsATAZygGsIQA9uii4CxOtAC8vfkJH4ipcgD4o00YokBteMggSAugG4AUKHoARCBDC6UGTCVXSSDZmy5QBC4lAD8UFY2dmhYJFp8gsIkBqoAXFAkpiYA9ClQAMoAlgC2YAhZ6FkQ7FAAbhAATpxZAqxewuYQALToEJzAUE0AdGbkUAAKcJXAWYhBYI7OiW4sHNzeYp0BAN4mUBtQWgNQWfWRcjF+CUMjYwgTjtuxpgC+UAlJJqkAVC-rL4kAFlncv1BwKBFBDQQh1YBwPZ7ADmUDBrA6cFYwG46AElQBCAQXS+0GAAjAUBBFWxkT+9QESAxHXREA+KRM8I6IgAcgBNAAaAGkAKJsjJqKBrTZQXlsgD6AEYEgBySKSmUUdabMXigBMssiasVyo2AGFUAAZQ08vUkACSqBZCWFIs2BuNpotVvFqulUBlYKxilqrC5-El4p1dvtRpNZstLNdfPVsq9IOIvv9IDVQaVItu6ZVMYAzJr+Dmdfc4NwmcBkk0RKwQIxk5wALJwQnSJpyK2csUZZIpN4fKAAJRslXabBRAKgHUqMKgSFYvoamOxqGrjCgrHROUQMhAnF6UBeDMrycFExCDlQOSywGwK7rjbAVBlDvDzpZMuUym7vf3A6HI+R3CApO06zvOciIEuK5wgI3qJnU267vSfT0HqMEJqMdTJgAQiWUjUHo563vwDZNloT5hk6kYyjczw9u8P6DmAw6cKOgFEleVRbsBrCwqB8HgViVY1tBsEYfsxF7p85qdBBAgAO6AWAkAcNOQHAFOPEAugLAYhAcCEF8InofOkQMN00LdFAAAGw5gOiwDipwAg5C0WTsFZkmHv0qGiUm-CClZAAkKw+cZmH8DhLG3MF3HQrcVlfvRnyMcxrHjrFM5zvxwgQUJq7rpUm7YkipTxj68Fkp5yHQMuNbHtIx4AD5QKF5V+vwFb9CE3DSLaGxaO23J8hk3RugYCSxaYIoDeyQ38qNMZquNa5IDkOBVFNmwzR2w3dM+lFWntFERi6Y02rqdpuRN6kwptIa-AAgpUBlZBU7AJDgAgwXprB3Xa7BwBC50hiGLmcJwcDQhA10adCf0g7snCDnA7DHFAn3fUi8ObLcf24xd21zSN+0nSyR2OqT0YSktwMg1SCAw7dF349Ng2dgtEo5strCretlR3M8R78CE95gNOvUEyQx57NucjdQC3BS-5TBzJ4rVwe1ICRXSCMBEFKxKyA0UrLF8UXSKjzJst3VaIbxgmPj1V5TyyKVCAji1YwITTCQnshFAzU81inX0J76tiR7K7e4Knsu+p7uDmClTsJgsVUL7Uc0NAgdIFiH7JOwRAIMMoJ1My9Y8iQAASqAWLTGzDijdQICAmSV7KLHADKf2N+wzetxX-YAOI8rKLmVFD3cXb3-dQIPI-iiTr5j1UUNlRrU8ijPrAty1JoPf2cYgsMU+O5WnsAKpgADLBqBdzV9SDtvS-UntW3E5sIwHQqf1-dp1DWesEBgBfAEO9Lo5A5AV2rrXboGRK7Yz-iKSIlsOq-yQRsMoWcEix1du7ZMIsmxix4s-fgNEMEZnQSGB+VCkEAMYEAkBYCEitmENAmuFhujzx5IgihKDEjJl4RgrBegTjDFGOMawkwCFZ1FjCUhIByEUI2CzO0twBpEUUffH+f8FG7HqOHPyWtcLLUfn-ehjDQHgNYXPSuHCuE8mHjyRex1XxCLtPww27iRQiJQAkRO6IU760NsbU2VkqCnAkRcKRmAZF6DkSQu2+daGqIzFoQx4VjEsXtiYQuhBi7DiBLODWM5r6AwgAAChvnAHBK4r43wgFoAwABKE4lRnK-AgJgMoAg3KfloukAA6qCJEUA8kFOgKM4YlQ4CtzbHUspLBuDrjkr0MsXgwCjEvB0LIhALCAxqXlep5SmmCi0BdMxmwLHAKsQkdhsD4EkCzCGfhcoAzBhBr46GHoZTM2eTokG1ymHgPuZw7h-zkH8HzCmD5IYvkJElH8i56CgW3NsTAsFjiR4Qs2Pw-WbNdpL0jOTF8kYqZSmNjKNyMp4o4swdgoU1SbS-GRtYyoKBbi3CRSKS5GxUXMPRfY7hLiKavjpduHBs12ZEsOjKqMbpxXwoBbo-FUrCWuOJXK8lkpKXUvigraCCJgCmKuj8zMtD+qqp2vNOVJKDrysWrq9gNKrIGrLKY+msoaVUDSAOSELFuBVHaRiXEhSKlJ2HMQZpVAcBIE6OuToMsQHQAekpPcIMuWUJyb64ZcJRlgBLNwS+izGkGC6AIcc6BiliTXBuCCIBui9CQCWipBItm-FGHsg5zTkjNoaRU85PKUUrksQK0FcDK7itefKWFdolUyl+Rmf5vKvAjpuWOuxsDwXoOnfwbUiqGWIqXcihG-KQWbsxU4qdULrLBQJTajVsrH0OolDq4KVLnW0vQUqlYTKhQsr0myjlmacbLuHYA9d56MUOKcSK0lVpr0gElda4mz67WUwVd+hlK7ppWqJuh18BGyVuidS6t1ZdjXMvAQu81GCtB4elWhrVqo1Skf1SWQ1HQPWVAZman16R+z+vaAwSowaoChugOG9EkbgDRvRnG2tib6jJqgKmsA6aQwgZUSYFpRggA)

This way is much more versatile than the previous one without losing it's type inference and type-safety, so I believe it will address all your concerns now.

The only issue that we have with this new way is shown in the playground (tomorrow @blazejkustra will help me test it in the App), but I believe that even with that issue it's worth using this new type because we can now use `OnyxUpdate` in the app to type our onyx update arrays.

* Example of a file where we define `optimisticData` -> https://github.com/Expensify/App/blob/main/src/libs/actions/IOU.js#L857
* We will also be able to type API methods to expect `OnyxUpdate[]` in `optimisticData`, `succesData` and  `failureData` inside `onyxData` -> https://github.com/Expensify/App/blob/main/src/libs/API.js#L40
* We can also type `pushData` to be `OnyxUpdate[]` and pass it directly to `Onyx.update()` -> https://github.com/Expensify/App/blob/main/src/libs/actions/User.js#L565

Reply:
That is very promising, let's test it together @fabioh8010!

Reply:
Update: we tested it and turns out it wasn't working on TS 4.8.4. We randomly tried a newer version of TS, 5.1.6 and it works like a charm! 🤩 

We are now testing if everything else work with TS 5.1.6, potentially we would like to update it in NewDot. Any thoughts? @hayata-suenaga @neil-marcellini